ipld/read.go

```diff
@@ -18,190 +18,162 @@ import (
+	// 2 and 4 to the right subtree
+	/* | 1 | 2 |
+	   | 3 | 4 |
+	*/
+	// choose sub-tree
+	if q%2 == 1 {
+		quadrant.isLeftSubtree = true
+	}
 
-// shareCounter is a thread safe tallying mechanism for share retrieval
-type shareCounter struct {
-	// all shares
-	shares map[index][]byte
-	// number of shares successfully collected
-	counter uint32
-	// the width of the extended data square
-	edsWidth uint32
-	// the minimum shares needed to repair the extended data square
-	minSharesNeeded uint32
+	// define range of shares for sampling
+	if q > 2 {
+		quadrant.from = edsWidth / 2
+		copy(quadrant.chunk, roots[edsWidth/2:])
+	} else {
+		copy(quadrant.chunk, roots[:edsWidth/2])
+	}
 
-	shareChan chan indexedShare
-	ctx       context.Context
-	cancel    context.CancelFunc
-	// any errors encountered when attempting to retrieve shares
-	errc chan error
+	return quadrant
 }
 
-func newshareCounter(parentCtx context.Context, edsWidth uint32) *shareCounter {
-	ctx, cancel := context.WithCancel(parentCtx)
+// fillQuadrant fetches 1/4 of shares for the given root
+func fillQuadrant(
+	ctx context.Context,
+	roots *quadrant, dag ipld.NodeGetter,
+	isRow bool,
+	dataSquare [][]byte,
+) error {
+	errGroup, ctx := errgroup.WithContext(ctx)
+	fetcher := func(i int) {
+		errGroup.Go(func() error {
+			rootHash, err := plugin.CidFromNamespacedSha256(roots.chunk[i])
+			if err != nil {
+				return err
+			}
+			subtreeRootHash, err := GetSubtreeLeaves(ctx, rootHash, dag, roots.isLeftSubtree)
+			if err != nil {
+				return err
+			}
 
-	// calculate the min number of shares needed to repair the square
-	minSharesNeeded := edsWidth * edsWidth / 4
+			leaves, err := GetLeaves(ctx, dag, subtreeRootHash, uint32(len(roots.chunk)))
+			if err != nil {
+				return err
+			}
 
-	return &shareCounter{
-		shares:          make(map[index][]byte),
-		edsWidth:        edsWidth,
-		minSharesNeeded: minSharesNeeded,
-		shareChan:       make(chan indexedShare, 512),
-		errc:            make(chan error, 1),
-		ctx:             ctx,
-		cancel:          cancel,
+			for leafIdx, leaf := range leaves {
+				shareData := leaf.RawData()[1:]
+				// it's not needed to store data for cols
+				// as we are fetching data from the same share for rows and cols
+				if isRow {
+					dataSquare[((i+roots.from)*len(roots.chunk)*2)+leafIdx] = shareData[NamespaceSize:]
+				}
+			}
+			return err
+		})
 	}
+
+	for i := 0; i < len(roots.chunk); i++ {
+		fmt.Println(i + roots.from)
+		fetcher(i)
+	}
+	return errGroup.Wait()
 }
 
-// retrieveLeaf uses GetLeafData to fetch a single leaf and counts that leaf
-func (sc *shareCounter) retrieveShare(
-	rootCid cid.Cid,
-	isRow bool,
-	axisIdx uint32,
-	idx uint32,
-	dag ipld.NodeGetter,
-) {
-	data, err := GetLeafData(sc.ctx, rootCid, idx, sc.edsWidth, dag)
+// GetSubtreeLeaves returns only one subtree - left or right
+func GetSubtreeLeaves(ctx context.Context, root cid.Cid, dag ipld.NodeGetter, isLeftSubtree bool) (cid.Cid, error) {
```

Comment:
Can you make this func to return the actual subtree by calling GetLeaves internally?

Reply:
moved to `GetSubtreeLeaves`